### DISCUSSION TITLE
[Paid] Precisely recreate color filter

### DISCUSSION THREAD

**Discussion by LeonardRockstar** [ImageMagick/ImageMagick#2631] [Help]
Hi! I'm trying to precisely precisely recreate a color filter that mostly adjusts the brightness of certain parts of an image. 
I tried using hald cluts to solve this, but the image can't be exported without some amount of artifacts, which leads to bad results. I also tried an algorithm (https://github.com/homm/color-filters-reconstruction), but the results weren't close enough. 

If anyone has experience with this sort of process, I'd appreciate it, and of course compensate you for your time.

**Comment by fmw42** [comment]
What is your IM version and platform/OS?  What was your exact command line when creating and using the HALD clut?  Did you use a lossy compressed image format? If so, that is bad. Does your filter do any convolutions or masking? If so, you cannot do that with HALD processing.  What does the filter do that you want to create. Is there some process or formula that you follow?

**Reply by LeonardRockstar** [reply]
Thanks a lot for the reply! I'm on Mac, the IM details are: "ImageMagick 7.0.10-30 Q16 x86_64 2020-09-21". I created the largest possible HALD clut that could be edited with the filter without scaling (14 in this case), then filtered and exported it.
I used PNG throughout. The filter does no masking or convolutions. The filter's only process in theory should be to darken highlight areas of the image and leaving the shadows intact, so it's easier to see overlaid text or other elements. 
However, it seems the filter does some sort of compression. The exported HALD isn't clean - I've attached a zoomed-in image. Do these artifacts look familiar to you? 

<img width="1136" alt="Bildschirmfoto 2020-09-26 um 13 29 13" src="https://user-images.githubusercontent.com/30780859/94340656-d7c75980-0003-11eb-9284-7464d0db7bcf.png">

I'm not really sure if could do something different to solve this, or if this requires some process of combining multiple images and eliminating the artifacts that way.

**Reply by fmw42** [reply]
What is your command line for processing the HALD image? To process the highlights and leave the shadows seems like processing that might be spatially dependent. If that is the case, HALD will not work. It only applies to global processing, i.e. the same thing is applied to every pixel.

**Reply by LeonardRockstar** [reply]
The filter only changes the bright areas of the image, that's what I was referring to, I would still apply it to the entire image. 
I've been experimenting with different variations (smaller hald images, rotating them, etc.), and have been able to generate hald images with very few artifacts. Applying it to photos looks good now, however, there's always a slight difference noticeable compared to filtering an image directly. Why could that be? I'm using PNG throughout. Could it be related to color profiles? I apply the hald image in terminal using image magick. My process is: Filtering hald image on phone -> exporting -> applying hald using image magick on a test image. Then, when I filter the test image directly, it looks slightly different. 

![comparison hald clut](https://user-images.githubusercontent.com/30780859/94352694-c82e2c00-0068-11eb-8217-49919b52f9b5.png)

The difference is very faint, there seems to be a blue shift in the highlight areas when I apply the hald.

**Reply by fmw42** [reply]
Does your input have a profile.  If you posted separate full resulting images, I could check.

But you can see how different they really are in the pixel data by using compare.

If on IM 7,

`magick compare -metric rmse hald_result direct_result null:`

If on IM 6, leave off the magick

`compare -metric rmse hald_result direct_result null:`

If the result is very small, then it is likely profiles. 

You can also see if your input or result has a profile by looking at the information from identify yourimage

Can you please show your command line for your enhancement, if it is not proprietary.

**Reply by LeonardRockstar** [reply]
Thanks so much for the help! I checked all the color profiles and properly converted the wrong ones in Photoshop, now my results are matching the filter.
